File: backend/src/main/java/ro/renovatorpro/domain/model/Item.java
package ro.renovatorpro.domain.model;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.Objects;

/**
 * Un element de cumpărat, aparținând unei camere (FK roomId). `quantity` poate fi fracționar (ex. mp).
 *
 * <p>{@code createdAt} — momentul adăugării, imutabil, setat o singură dată la creare.
 * {@code purchasedAt} — momentul ultimei tranziții spre {@link ItemStatus#CUMPARAT} (null dacă elementul
 * nu a fost niciodată cumpărat); folosit de graficul „Evoluția Cheltuielilor" (Problema 3 din audit) —
 * NU se resetează dacă statusul revine la altceva, dar se actualizează la o nouă tranziție spre Cumpărat.
 */
public record Item(
        String id,
        String roomId,
        String name,
        MaterialType materialType,
        String source,
        ItemStatus status,
        BigDecimal quantity,
        Money unitPrice,
        String productUrl,
        String imageUrl,
        ItemOrigin origin,
        Instant createdAt,
        Instant purchasedAt
) {

    public Item {
        Objects.requireNonNull(id, "id");
        Objects.requireNonNull(roomId, "roomId");
        Objects.requireNonNull(name, "name");
        Objects.requireNonNull(materialType, "materialType");
        Objects.requireNonNull(status, "status");
        Objects.requireNonNull(quantity, "quantity");
        Objects.requireNonNull(unitPrice, "unitPrice");
        Objects.requireNonNull(origin, "origin");
        Objects.requireNonNull(createdAt, "createdAt");
        if (quantity.signum() < 0) {
            throw new IllegalArgumentException("Cantitatea nu poate fi negativă: " + quantity);
        }
    }
}

File: frontend/src/shared/types/Item.ts
import { ItemOrigin } from "./ItemOrigin";
import { ItemStatus } from "./ItemStatus";
import { MaterialType } from "./MaterialType";

/** Un element de cumpărat, aparținând unei camere (FK roomId). */
export interface Item {
  id: string;
  roomId: string;
  name: string;
  materialType: MaterialType;
  source: string;
  status: ItemStatus;
  quantity: number;
  unitPrice: number;
  productUrl?: string;
  imageUrl?: string;
  /** Manual (adăugat de user) sau Configurare (generat automat din pardoseală/plintă/faianță). */
  origin: ItemOrigin;
  /** Momentul adăugării (ISO 8601), setat de server — imutabil, nu editabil de user. */
  createdAt: string;
  /**
   * Momentul ultimei tranziții spre `ItemStatus.Cumparat` (ISO 8601), setat de server — absent dacă
   * elementul nu a fost niciodată cumpărat. Folosit de graficul „Evoluția Cheltuielilor" (Problema 3).
   */
  purchasedAt?: string;
}
